src/provider/mod.rs

```diff
@@ -243,6 +261,244 @@ where
+/// which is used by the associated functions.
+pub struct L1TransactionReceipt<T> {
+    /// Ethereum transaction receipt.
+    inner: TransactionReceipt,
+    /// A reference to the L2 provider.
+    l2_provider: RootProvider<T, Zksync>,
+}
+
+impl<T> L1TransactionReceipt<T>
+where
+    T: Transport + Clone,
+{
+    pub fn new(tx_receipt: TransactionReceipt, l2_provider: RootProvider<T, Zksync>) -> Self {
+        Self {
+            inner: tx_receipt,
+            l2_provider,
+        }
+    }
+
+    pub fn get_receipt(&self) -> &TransactionReceipt {
+        &self.inner
+    }
+
+    pub fn get_l2_tx(&self) -> Result<PendingTransactionBuilder<T, Zksync>, L1CommunicationError> {
+        let l1_to_l2_tx_log = self
+            .inner
+            .inner
+            .logs()
+            .iter()
+            .filter_map(|log| log.log_decode::<Bridgehub::NewPriorityRequest>().ok())
+            .next()
+            .ok_or(L1CommunicationError::NewPriorityRequestLogNotFound)?;
+
+        let l2_tx_hash = l1_to_l2_tx_log.inner.txHash;
+
+        Ok(PendingTransactionBuilder::new(
+            self.l2_provider.clone(),
+            l2_tx_hash,
+        ))
+    }
+}
+
+/// Trait for ZKsync provider with populated wallet
+/// Contains provider methods that need a wallet
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait ZksyncProviderWithWallet<T = BoxTransport>:
+    ZksyncProvider<T> + WalletProvider<Zksync>
+where
+    T: Transport + Clone,
+{
+    /// Deposits specified L1 token to the L2 address.
+    ///
+    /// ## Parameters
+    ///
+    /// - `token_address`: L1 token address to deposit.
+    /// - `receiver`: receiver L2 address.
+    /// - `amount`: amount to deposit in Wei.
+    /// - `l1_provider`: reference to the L1 provider.
+    ///
+    /// ## Returns
+    ///
+    /// L1TransactionReceipt.
+    /// Hint: use returned L1 transaction receipt to get corresponding L2 transaction and wait for its receipt
+    /// E.g.: deposit_l1_receipt.get_l2_tx()?.with_required_confirmations(1).with_timeout(Some(std::time::Duration::from_secs(60 * 5))).get_receipt()
+    async fn deposit<Tr, P>(
+        &self,
+        token_address: Address,
+        receiver: Address,
+        amount: U256,
+        l1_provider: &P,
+    ) -> Result<L1TransactionReceipt<T>, L1CommunicationError>
+    where
+        Tr: Transport + Clone,
+        P: alloy::providers::Provider<Tr, Ethereum>,
+    {
+        if token_address != ETHER_L1_ADDRESS {
+            return Err(L1CommunicationError::Custom(
+                "Deposit for the specified token is not currently supported.",
+            ));
+        }
+        let l2_chain_id = self.get_chain_id().await.map_err(|_| {
+            L1CommunicationError::Custom("Error occurred while fetching L2 chain id.")
+        })?;
+        let sender = self.wallet().default_signer_address();
```

Comment:
We will probably have to fix that in the future, so that users can choose which wallet must be used for the deposit, but it's OK for now.